File: docs-examples/example-kotlin/src/test/kotlin/io/micronaut/nats/docs/headers/ProductClient.kt
package io.micronaut.nats.docs.headers

import io.micronaut.context.annotation.Requires
// tag::imports[]
import io.micronaut.messaging.annotation.MessageBody
import io.micronaut.messaging.annotation.MessageHeader
import io.micronaut.messaging.annotation.MessageHeaders
import io.micronaut.nats.annotation.NatsClient
import io.micronaut.nats.annotation.Subject
import io.nats.client.impl.Headers
// end::imports[]

@Requires(property = "spec.name", value = "HeadersSpec")
// tag::clazz[]
@NatsClient
@MessageHeaders(
    MessageHeader(name = "x-product-sealed", value = "true"), // <1>
    MessageHeader(name = "productSize", value = "large")
)
interface ProductClient {

    @Subject("product")
    @MessageHeaders(
        MessageHeader(name = "x-product-count", value = "10"), // <2>
        MessageHeader(name = "productSize", value = "small")
    )
    fun send(data: ByteArray)

    @Subject("product")
    fun send(@MessageHeader productSize: String?, // <3>
             @MessageHeader("x-product-count") count: Long,
             data: ByteArray)

    @Subject("products")
    @MessageHeader(name = "x-product-count", value = "20")
    fun send(@MessageBody data:ByteArray, @MessageHeader productSizes: List<String>) // <4>

    @Subject("productHeader")
    fun send(@MessageBody data: ByteArray, headers: Headers) // <5>

}
// end::clazz[]

File: docs-examples/example-python/src/test/python/micronaut/nats/docs/headers/ProductClient.py
from abc import ABC, abstractmethod
from typing import Annotated

from java.lang import Long
from micronaut.context.annotation import Requires
try:
    # tag::imports[]
    from micronaut.messaging.annotation import MessageBody, MessageHeader
    from micronaut.nats.annotation import NatsClient, Subject
    from io.nats.client.impl import Headers
    # end::imports[]
except ImportError:  # TODO(python): packages under `io.` other than `io.micronaut` cannot be imported at runtime
    from nats.client.impl import Headers


@Requires(property="spec.name", value="HeadersSpec")
# tag::clazz[]
@NatsClient
@MessageHeader(name="x-product-sealed", value="true")  # <1>
@MessageHeader(name="productSize", value="large")
class ProductClient(ABC):

    @Subject("product")
    @MessageHeader(name="x-product-count", value="10")  # <2>
    @MessageHeader(name="productSize", value="small")
    @abstractmethod
    def send(self, data: bytes) -> None:
        ...

    @Subject("product")
    @abstractmethod
    def send_with_headers(self, product_size: Annotated[str | None, MessageHeader("productSize")],  # <3>
                          count: Annotated[Long, MessageHeader("x-product-count")],
                          data: bytes) -> None:
        ...

    @Subject("products")
    @MessageHeader(name="x-product-count", value="20")
    @abstractmethod
    def send_with_product_sizes(self, data: Annotated[bytes, MessageBody],
                                product_sizes: Annotated[list[str], MessageHeader("productSizes")]) -> None:  # <4>
        ...

    @Subject("productHeader")
    @abstractmethod
    def send_with_nats_headers(self, data: Annotated[bytes, MessageBody], headers: Headers) -> None:  # <5>
        ...
# end::clazz[]
